### ISSUE TITLE
Testing kinetic lancer

### ISSUE THREAD

**Issue by NosVladimir** [NosVladimir/KineticArchetypes#8]
Hi @deng-tianhan, I've finished the lancer archetype.

Since you are quite active in playing with this mod, may I ask if you are happy to test it? Most of the lancer features and abilities are implemented with my (potentially unstable) custom code, so it's likely that my personal testing does not cover enough edge cases.

If you are happy to test, you can find a newest v1.0.0 release in the releases page.

**Comment by deng-tianhan** [comment]
Minor bug: locked offhand

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/56fd027d-f821-48c9-9a8b-9221235ff3ce)

Steps to reproduce:

1. Equip kinetic lance (simple blast)
2. Equip kinetic blade (composite blast, or by clicking on a different element)
3. Unequip kinetic blade

Locked offhand makes the character unable to equip shield or dual wield. Two handers can still be equipped though.

Steps to fix it:

1. Equip kinetic lance
2. Unequip kinetic lance
3. Offhand becomes unlocked again

**Comment by NosVladimir** [comment]
Ohh thanks for the report, this will be fixed in the next update (formal release)

**Comment by deng-tianhan** [comment]
Diving into combat only uses up standard action.

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/efba6d93-320d-42f1-a983-0195c9867bf2)
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/5d5b9e07-462c-4ab1-bacb-0becd255bb39)

**Comment by NosVladimir** [comment]
Hmm... For me dragoon diving into combat will automatically attack and end the turn (in turn based).

Theoretically it should behave like charging into combat. Could you double check with a normal charge?

**Comment by deng-tianhan** [comment]
Missing damage source.

But I'm not sure if this is from impaling crash? Seems like normally the damage number appears before the line with "strength check failed", so I suspect this damage is something unrelated. (Vrok being mythic and on its 2nd life?)

Normal:
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/d553976f-28a0-4103-85aa-7fc4e0af798e)

Missing damage source:
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/2dbf84fd-fee1-4462-bdf6-f600f7b6832c)

**Comment by deng-tianhan** [comment]
> Hmm... For me dragoon diving into combat will automatically attack and end the turn (in turn based).

I tried a few times with normal charge and dive and seems like no issue.

I suspect it was the surprise round ending immediately, which led me to think I still had move action by mistake. In fact the move action should be from the normal round after surprise round

**Comment by NosVladimir** [comment]
Wow, damage out of nowhere. I'm pretty sure it's not directly due to impaling crash, but could be indirectly. Does the character have any mythic abilities or any other buffs that grant additional damage? E.g. sword of heaven with specific selections

**Comment by deng-tianhan** [comment]
> Does the character have any mythic abilities or any other buffs that grant additional damage?

Double checked no buffs and haven't taken mythic level yet.

I suspect the mythic vrock's hard to kill effect? Non mythic enemies seem fine

**Comment by deng-tianhan** [comment]
> that's normal and I'm unable to fix unless I make the already complicated code 3 times more complicated to factor in moving targets and live-updating a jump curve...

I think the current is quite good and no need to make major changes like that 😅 

I like the jumping animation btw, it's very different from what I expected (in a good way). Because dragoon and diving makes me think of Final Fantasy dragoons where they just jump up / down vertically instead of moving in a curve... IMHO your version feels to fit the PF theme a lot more 😄

**Comment by deng-tianhan** [comment]
I'm not sure if it's a reference but dragoons are definitely one of the iconic classes / characters from FF 😄 

some examples of their jump
https://www.youtube.com/watch?v=Pzgq1248CV8

**Comment by deng-tianhan** [comment]
Lancer can get working maximize from metakinetic master

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/2050c141-6858-4779-bdf9-5eef98e00dfb)

This is a bit of grey area, since impaling crush says `replaces metakinesis (maximize)`, but metakinetic master never mentions requiring maximize... Personally I like to have maximize, but I suspect DMs would rule otherwise. 🤣

**Comment by NosVladimir** [comment]
> I'm not sure if it's a reference

Maybe, maybe not, coz the archetype comes from a third party book and it's relatively new

After seeing the example I realised I've seen the dragoon jump from the FFXIV crossover in MHW. That truly would be cool to implement in WOTR, but let me keep the dragoon dive as it is now

> Lancer can get working maximize from metakinetic master

I forgot to add a no-archetype prerequisite to metakinesis maximized in the metakinesis master. As it should read, 'a kineticist chooses one type of metakinesis and reduces the burn cost', so even if a lancer can choose maximized, they don't and shoundn't have the base ability to apply it to.

Owlcat, on the other hand, implemented metakinesis master as three separate base abilities that would replace original ones upon selection, so that bypasses the rule.

The same would apply to a kinetic duelist, I also didn't add the restriction to metakinesis quicken. These will be properly restricted in the next update

**Comment by deng-tianhan** [comment]
I think the current dive is pretty good as is 😄 

Just an idea: if you want to implement the vertical jumping later on, maybe can keep both version and add a toggle to let player decide?

**Comment by deng-tianhan** [comment]
On a separate topic - I keep seeing vital blade being mentioned on d20PFSRD and other places but couldn't find a source. It's not in game either, so is it from third party?

**Comment by NosVladimir** [comment]
Yes, it's vital strike with kinetic blades, as an infusion. https://libraryofmetzofitz.fandom.com/wiki/Kinetic_Blade#VITAL_BLADE

**Comment by NosVladimir** [comment]
Fixed the offhand locked bug and restricted the metakinesis selections. zip file updated under v1.0.0

**Comment by deng-tianhan** [comment]
When surprise round starts normally (not triggered by dive or charge), charge can be used as a standard action. But dive cannot be selected (1st slot in action bar in the image below)

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/f0fb64b1-3057-4928-8ffb-d9cbc1292bc1)

**Comment by deng-tianhan** [comment]
Not sure if intended: I noticed dive is useable while fatigued. Charge is not.

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/0fe1a98d-2bef-446a-b690-0f7313f0ca13)

**Comment by deng-tianhan** [comment]
If a trap is located between character position and the jump target position, jumping triggers the trap. (As if walked across the trap)

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/5ff20a17-6dd7-4a80-9380-616c05ea534d)

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/b3321e1d-9584-4341-ab9a-2670a43ebe17)

**Comment by deng-tianhan** [comment]
This one is hilarious🤣 

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/6f0b72fa-d097-497f-b56b-0f79fef810ea)

While indoors, jump can actually target the top of a wall (black area)

Once the character is standing on the wall, jump doesn't work at all (cannot target the floor or the wall area). But the character is able to simply walk from the top of the wall to the nearby floor.

**Comment by deng-tianhan** [comment]
During turn based mode, when the lancer is outside of the screen and use dive to target an enemy inside the screen, the charge animation is played instead.

**Comment by NosVladimir** [comment]
> charge can be used as a standard action. But dive cannot

I didn't figure out how charge can be used in surprise round. I think it's fine that dive can't be used in surprise round because it says 'as a full round action', which is mechanically different from a charge despite it giving a charge buff

> I noticed dive is useable while fatigued. Charge is not.

The rule doesn't say anything about jumping while fatigued, but I agree it makes sense to enforce the restriction. Among other restrictions of a charge, entangled and exhausted also make sense for jump/dive. Difficult terrain and corruption level 2 debuff I don't think shall apply to jump

> jumping triggers the trap

This I can fix by making the character untargetable during a jump, but need to test if some edge cases would result in the character permanently untargetable...

> dive to target an enemy inside the screen, the charge animation is played instead

Didn't expect that, but there's nothing I can do

**Comment by NosVladimir** [comment]
Updated the restrictions and made character untargetable during the jump

**Comment by NosVladimir** [comment]
That's intended. IMO to start a jump does not have any difficulty as such, but to properly land at a desired distance without falling does. You have to know your (characters') limits and not stregtch them too far in combat, otherwise you do take the risk.

A natural 1 for a skill check does not make it automatically fail, so as long as you know roughly how much all the bonuses add up to, you are certain to succeed a check in jump within a corresponding horizontal distance. Vertical height is indeed harder to estimate, but it's rare to need a very high jump - again, if that is needed, take the risk or don't jump

**Comment by deng-tianhan** [comment]
Right, agree that player should know the character's limit.

What I wanted to highlight is it creates a somewhat funny effect outside of combat: player can attempt a DC 50 jump on purpose and fall flat, but the character gets a free "teleport" effect.

**Comment by NosVladimir** [comment]
A long jump would be the same as increments of shorter jumps unless it's over a gap or something dangerous

Do you mean the jump can be used on purpose to reach some weird places like can be done with a dimensional door but for free? I think it's okay - jump can only be casted on destinations the character can see, unless the line of sight detection is not working

**Comment by deng-tianhan** [comment]
> Do you mean the jump can be used on purpose to reach some weird places like can be done with a dimensional door but for free?

Yup. Having fun exploring weird places now 🤣 

line of sight is working, so it's really not an issue

**Comment by deng-tianhan** [comment]
BTW, I'm happy to see that equipment with on charge special effects are triggering on dive.

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/a8932a0b-c181-45b6-addf-ca7817eae65f)

**Comment by NosVladimir** [comment]
How the dual and the spear abilities work is to apply a corresponding buff when used, then when using/equipping the kinetic blades, corresponding changes are made. When deactivating the blade, the buffs also disappear, so if players ever want to equip a normal kinetic blade they can still do that.

I guess the game deactivates all activatables and re-activate them when the character levels up, which would reset the blades.

The dual blade ability is among my very first modding attempts so I just used a simple one-off ability rather than an activatable. 
Made spear in the same fashion to save time, but looking back I agree they should be activatables that don't go off unless manually switched off. I'll try to re-make these abilities into activatables.

**Comment by deng-tianhan** [comment]
BTW I think one of the PFKM mods implemented proper reach weapon: the reach weapon can only attack at 10ft but not 5ft. So when enemy walks from 10ft to 5ft you get AOO.

I wish the same option is added to WotR, combined with mythic combat reflex:

1. Kinetic lancer dive to kill enemy spellcasters
2. AOO on enemy archers because still within reach from lance + whip + lunge
3. AOO on enemy melees who walk towards the lancer 😄

**Comment by NosVladimir** [comment]
Interesting thought, but according to RAW kinetic spear works the same way as whip ("Unlike most reach weapons, the kinetic whip can also attack nearby creatures. "), except for the crit range increase, plus the drawback of reduced damage dice and can't crit within natural reach.

For simplicity I didn't implement the latter half of the spear effect, also didn't restrict its usage with whip from DarkCodex... Maybe we can formulate this as a pole with a whip in the front replacing a spearhead, so it's extra extra reach

**Comment by NosVladimir** [comment]
Okay I've made dual and spear into activatables and updated the v1.0.0 zip. Hopefully that wouldn't break any previous saves - @deng-tianhan could you confirm that?

**Comment by deng-tianhan** [comment]
Cool! Mostly no issue.

Here's the detailed effect on existing saves:

1. character still holding lance on load
2. new ability is not activated on load
3. old toggle disappeared from quick action bar. new ability can be found under abilities drawer
4. activate ability - no effect as character is already holding lance
5. deactivate ability - lance change into blade

Same pattern for dual blades.

**Comment by deng-tianhan** [comment]
Interestingly deactivating lance first doesn't change reach. Deactivating whip first changes reach.

lance + whip + lunge
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/65e0d2d8-5a27-4b5a-a695-67a062f787f0)

deactivate lance only
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/c6db38b7-db35-44cb-b4d5-213459fc65d4)

deactivate whip only
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/1cec23f5-9a60-4381-a47c-e79554455413)

**Comment by deng-tianhan** [comment]
Fun fact: free AOO when enemy leave your reach. This works when you change size, etc.

1. Use kinetic lance or esoteric blade (allows AOO)
2. Activate whip
3. Move towards the enemy until they are within reach
4. Deactivate whip
5. AOO

IMO no need to fix. 😋

**Comment by NosVladimir** [comment]
> Deactivate whip
AOO

Can you do that infinitely in a round? I'm at work atm so can't check, IIRC whip is a free action? Maybe you can report this to Truinto

**Comment by deng-tianhan** [comment]
> Can you do that infinitely in a round?

Yes! 🤣 

Deactivate lance can replicate the same. So esoteric blade + lance should be able to pull it off as well, even without whip.

**Comment by NosVladimir** [comment]
That is definitely a game-breaking synergy... Esoteric blade + lance might not work, because lance/spear needs to sheath the blade then re-draw it, and I'm pretty sure sheathing doesn't let you AoO enemies (I never tried) - unless you can AoO with simply swapping out to another weapon set.

I can do a simple fix by changing whip to also deactivate in the next round, this can be annoying if people do want to turn off whip immediately for whatever reason. An alternative fix would be to make whip also re-equip kinetic blade, and hopefully we don't get free AoOs

**Comment by deng-tianhan** [comment]
I think if lance and whip follow lunge implementation it should resolve the spammable issue.

You might still get 1 free AOO at the start of the next round, but that's more of a base game thing. (e.g. enlarge expire)

**Comment by NosVladimir** [comment]
That sounds like a good plan. Do you have toybox enabled? If you do, you can try setting the 'Immediately deactivate activatables' option (I think it has one under Cheats, may not be named exactly like this) and see if the same happens for lunge

**Comment by deng-tianhan** [comment]
> Esoteric blade + lance might not work, because lance/spear needs to sheath the blade then re-draw it

I think it should still work. I tried keeping whip on, toggling lance gives free AOO.
Esoteric blade version is just using it for AOO capability when lance is turned off.

During combat the character just change from holding spear to holding sword animation. There's no drawing weapon animation.
Perhaps due to kinetic spear is still kinetic blade?

**Comment by deng-tianhan** [comment]
Hmm... can't seem to find the option. Maybe not under cheats section?

![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/abbfe02c-d362-49f1-ac76-1e2cbeacbcb5)

**Comment by NosVladimir** [comment]
I do remember there is an option, maybe not under Cheats. No worries if you can't find it, I can also test that

**Comment by deng-tianhan** [comment]
I tried the party option to remove the lunge buff, and the AOO triggered. 🤣

**Comment by deng-tianhan** [comment]
> An alternative fix would be to make whip also re-equip kinetic blade

That might be too hard to use during combat because the game still eats my move (and then standard) actions even though (un)equipping kinetic blade is shown as free in the action bar.

e.g. when I don't have 0 burn whip, and only want to use it when enemy is far. re-equip approach will not work well.

**Comment by NosVladimir** [comment]
> That might be too hard to use during combat

Yes, and also from your feedback, lance despite already using that mechanism, still has the problem of spammable AoOs, so I probably will stick to delay the deactivation

**Comment by deng-tianhan** [comment]
Sure, delay the deactivation would be the lesser of the evils.

Players will have to be more careful about how much more burn they can take and still keep 0 burn lance. Once the character reaches higher levels then this becomes non-issue.

**Comment by deng-tianhan** [comment]
> lance despite already using that mechanism, still has the problem

Oh I'm already using the build with activatable lance. Buff version lance shouldn't have this issue.

But esoteric blade + whip would still get spammable AOOs.

**Comment by deng-tianhan** [comment]
BTW I enjoy the trap fix a lot. Jumping to the other side and then disabling it is very stylish.

Not sure if you can use it to dodge spells in real time with pause... but who plays that mode, right? I always play turn based. 😝

**Comment by deng-tianhan** [comment]
I thought of some test cases for your reference:

- [x] base blade + lance. not spammable
- [x] base blade + whip. not spammable
- [x] lance + whip. toggle lance. spammable
- [x] lance + whip. toggle whip. spammable
- [ ] esoteric blade + lance
- [ ] esoteric blade + whip
- [ ] esoteric blade + lance + whip

**Comment by NosVladimir** [comment]
I tested the interactions and for all the cases where you got spammable AoOs, I only got 1 AoO (made sure to have combat reflexes and 100 dex modifier). I also tried with repeated toybox teleporting in/out of attack range and the AoO also only happens once. I guess it's another mod that made it so that AoOs caused by leaving the attack range can happen unlimitedly in a round? Or it might be my testing target enemy and/or my testing character is a bit bugged.

Anyway, I've changed spear and whip to deactivate in the next round, so they should no longer be spammable to get AoOs. The immediate deactivate option is actually from DarkCodex instead of toybox.

> Not sure if you can use it to dodge spells in real time with pause

In principle yes. It would be very cool to do so in real time mode.

> but who plays that mode, right?

Actually some mod devs only play RTwP, and it's a war every time the debate starts... So be careful when you want to mention this under other mods XD

I see you are also using impaling crash, have you encountered any issues for it? I know in turn based sometimes the damage triggers in one round later, but it's a base game issue with how asynchronous each unit's time is in turn based and I can't fix that.

If things are ok, I plan to publish a formal release this weekend. I've also passed the mod to pheonix99 (author of mod TomeOfTheFirebird) since she had kinetic lancer on her list and kindly provided me with some initial code for the jump, and she's also happy about how dragoon dive works.

**Comment by deng-tianhan** [comment]
> I also tried with repeated toybox teleporting in/out of attack range and the AoO also only happens once. I guess it's another mod that made it so that AoOs caused by leaving the attack range can happen unlimitedly in a round?

Good catch! Only 1 AOO from leaving attack range sounds right. I didn't notice initially because my character was one-shotting test targets.

I used toybox to force roll 1s, and verified that I only got 1 AOO per enemy. So it's still unlimited AOOs for me but capped by number of enemies.

> Anyway, I've changed spear and whip to deactivate in the next round

I will test with this build to see in a bit.

> Actually some mod devs only play RTwP, and it's a war every time the debate starts...

I was joking 😝 It's just personal preference and I don't intend to change other people's opinion

> impaling crash, have you encountered any issues for it? I know in turn based sometimes the damage triggers in one round later

So far no issues. The damage triggering later is most noticeable when diving into combat:

1. Dive into combat (surprise round)
2. Dive again when battle starts (high initiative roll)
3. Damage ticks in enemy round.

I'm fine with the current behavior. Though I can see some people might prefer damage ticking in the dragoon's round?

> If things are ok, I plan to publish a formal release this weekend. I've also passed the mod to pheonix99 (author of mod TomeOfTheFirebird) since she had kinetic lancer on her list and kindly provided me with some initial code for the jump, and she's also happy about how dragoon dive works.

Great! Looking forward to the formal release (and potentially more collaboration between mod authors?) 😄

**Comment by deng-tianhan** [comment]
Tested delay deactivation fixed the issue

Before: free AOOs (1 per enemy)
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/62856ecb-4636-4def-97d0-07c6eddd3636)

After: reach change delayed. Character has same reach when deactivate lance, whip, lunge
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/cb50c74b-533b-4598-acbc-c066b41a8afb)

**Comment by deng-tianhan** [comment]
The spear seems to get unequipped when changing map sometimes? I can't reproduce it consistently though, maybe only happens to random encounters.

Just for reference, I didn't notice this before changing spear to activatable.

**Comment by deng-tianhan** [comment]
In some rare cases, the lancer doesn't attack after a dive with a successful mobility check. Not sure why though

No attack after the dive in combat log
![image](https://github.com/NosVladimir/KineticArchetypes/assets/24478731/7f1afa8c-cc22-4837-bebd-e5d91f1612c3)


>[1046.0720 - TechArt]: Camera at the end: (38.2, 0.1, -9.4)
[1050.1087 - TechArt]: Camera at the end: (27.2, 0.1, -32.2)
[1061.6091 - Mods]: KineticArchetypes.KineticLancer: Leap from (27.1, 0.1, -32.2) to [Target: unit 'Unit[CR3_AbrikandiluStandard]#2F695']
[1062.0025 - History.Skill]: 03:34:02 20.08.4715 - [History.Skill] StartGame_Player_Unit (4391e8b9afbb0cf43aeba700c089f56d)SkillMobility check passed | dc = 23 | result = 86 | d20 = 9 | stat value = 32
[1084.3800 - System]: Start mode request EscMode
[1084.3960 - System]: Started mode EscMode (previous mode Default)

**Comment by NosVladimir** [comment]
> The spear seems to get unequipped when changing map sometimes? (similar to kinetic blade removed after rest)
I can't reproduce it consistently though, maybe only happens to random encounters.

I have no idea with these two... Added a line to let it re-activate when map loads, hopefully that will have some effect

> In some rare cases, the lancer doesn't attack after a dive with a successful mobility check. Not sure why though

It's normal if it happens when diving into combat - if the enemy is moving fast enough, by the time you finish jumping to the destination, the enemy may have left your reach so you can't attack. If it ever happens in other situations, then it might be a base game bug since I used the same attack logic as a charge.

**Comment by NosVladimir** [comment]
Made a formal release. Thank you for your patience in testing!! I'll close this issue now and lock its conversations. 

If you find new bugs, just create a new issue.
